File: staff-engagement-frontend/src/app/portfolios/portfolio.service.ts
import { inject, Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';

import { environment } from '../../environments/environment';
import {
  CreateEducationRequest,
  CreateProjectRequest,
  CreateShowcaseLinkRequest,
  EducationResponse,
  PortfolioResponse,
  ProjectResponse,
  ShowcaseLinkResponse,
  UpdateEducationRequest,
  UpdateProjectRequest,
  UpdateShowcaseLinkRequest
} from './portfolio.models';

@Injectable({ providedIn: 'root' })
export class PortfolioService {
  private readonly http = inject(HttpClient);

  private baseUrl(employeeId: number): string {
    return `${environment.apiUrl}/employees/${employeeId}/portfolio`;
  }

  getPortfolio(employeeId: number): Observable<PortfolioResponse> {
    return this.http.get<PortfolioResponse>(this.baseUrl(employeeId));
  }

  // Education

  addEducation(employeeId: number, request: CreateEducationRequest): Observable<EducationResponse> {
    return this.http.post<EducationResponse>(`${this.baseUrl(employeeId)}/education`, request);
  }

  updateEducation(employeeId: number, id: number, request: UpdateEducationRequest): Observable<EducationResponse> {
    return this.http.put<EducationResponse>(`${this.baseUrl(employeeId)}/education/${id}`, request);
  }

  deleteEducation(employeeId: number, id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl(employeeId)}/education/${id}`);
  }

  // Projects

  addProject(employeeId: number, request: CreateProjectRequest): Observable<ProjectResponse> {
    return this.http.post<ProjectResponse>(`${this.baseUrl(employeeId)}/projects`, request);
  }

  updateProject(employeeId: number, id: number, request: UpdateProjectRequest): Observable<ProjectResponse> {
    return this.http.put<ProjectResponse>(`${this.baseUrl(employeeId)}/projects/${id}`, request);
  }

  deleteProject(employeeId: number, id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl(employeeId)}/projects/${id}`);
  }

  // Showcase links

  addLink(employeeId: number, request: CreateShowcaseLinkRequest): Observable<ShowcaseLinkResponse> {
    return this.http.post<ShowcaseLinkResponse>(`${this.baseUrl(employeeId)}/links`, request);
  }

  updateLink(employeeId: number, id: number, request: UpdateShowcaseLinkRequest): Observable<ShowcaseLinkResponse> {
    return this.http.put<ShowcaseLinkResponse>(`${this.baseUrl(employeeId)}/links/${id}`, request);
  }

  deleteLink(employeeId: number, id: number): Observable<void> {
    return this.http.delete<void>(`${this.baseUrl(employeeId)}/links/${id}`);
  }
}

File: staff-engagement-backend/src/main/java/com/psybergate/staff_engagement/portfolio/PortfolioService.java
package com.psybergate.staff_engagement.portfolio;

import com.psybergate.staff_engagement.portfolio.dto.CreateEducationRequest;
import com.psybergate.staff_engagement.portfolio.dto.CreateProjectRequest;
import com.psybergate.staff_engagement.portfolio.dto.CreateShowcaseLinkRequest;
import com.psybergate.staff_engagement.portfolio.dto.EducationResponse;
import com.psybergate.staff_engagement.portfolio.dto.PortfolioResponse;
import com.psybergate.staff_engagement.portfolio.dto.ProjectResponse;
import com.psybergate.staff_engagement.portfolio.dto.ShowcaseLinkResponse;
import com.psybergate.staff_engagement.portfolio.dto.UpdateEducationRequest;
import com.psybergate.staff_engagement.portfolio.dto.UpdateProjectRequest;
import com.psybergate.staff_engagement.portfolio.dto.UpdateShowcaseLinkRequest;

/**
 * Module boundary for the portfolio module.
 *
 * <p>Cross-module callers (notably the future skills module) depend on this interface, never on
 * the portfolio repositories, keeping the modular monolith splittable.
 */
public interface PortfolioService {

    PortfolioResponse getPortfolio(Long employeeId);

    EducationResponse addEducation(Long employeeId, CreateEducationRequest request);

    EducationResponse updateEducation(Long employeeId, Long educationId, UpdateEducationRequest request);

    void deleteEducation(Long employeeId, Long educationId);

    ProjectResponse addProject(Long employeeId, CreateProjectRequest request);

    ProjectResponse updateProject(Long employeeId, Long projectId, UpdateProjectRequest request);

    void deleteProject(Long employeeId, Long projectId);

    ShowcaseLinkResponse addLink(Long employeeId, CreateShowcaseLinkRequest request);

    ShowcaseLinkResponse updateLink(Long employeeId, Long linkId, UpdateShowcaseLinkRequest request);

    void deleteLink(Long employeeId, Long linkId);

    boolean projectExists(Long projectId);
}
